File: lab_7_3.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int MAX_RECURSION_DEPTH = 1000; // Максимальная глубина рекурсии
int recursion_depth = 0; // Счётчик глубины рекурсии

bool canPlaceItems(vector<int>& items, vector<int>& containers, int H, int index) {
    // Проверка на превышение глубины рекурсии
    if (recursion_depth > MAX_RECURSION_DEPTH) {
        cout << "Превышена максимальная глубина рекурсии!" << endl;
        return false;
    }
    recursion_depth++;
    
    // Все предметы размещены успешно
    if (index == items.size()) {
        return true;
    }
    
    // Пробуем разместить текущий предмет в каждом контейнере
    for (int i = 0; i < containers.size(); i++) {
        if (containers[i] + items[index] <= H) { // если по высоте проходит
            containers[i] += items[index];
            if (canPlaceItems(items, containers, H, index + 1)) {
                return true;
            }
            // Откатываем изменения 
            containers[i] -= items[index]; // заново
        }
    }
    
    return false;
}

int main() {
    int N, H, M;
    cout << "Введите количество контейнеров (N): ";
    cin >> N;
    cout << "Введите высоту каждого контейнера (H): ";
    cin >> H;
    cout << "Введите количество предметов: ";
    cin >> M;
    
    vector<int> items(M);
    cout << "Введите высоты предметов: ";
    for (int i = 0; i < M; i++) {
        cin >> items[i]; // для каждого
    }
    
    // Сортируем предметы в порядке убывания для оптимизации
    sort(items.begin(), items.end(), greater<int>());
    
    vector<int> containers(N, 0); // Изначально все контейнеры пусты
    
    recursion_depth = 0; // Сброс счётчика перед вызовом
    if (canPlaceItems(items, containers, H, 0)) {
        cout << "Предметы можно разместить в контейнерах!" << endl;
    } else {
        cout << "Предметы нельзя разместить в контейнерах." << endl;
    }
    
    return 0;
}

File: lab_7_3.rb
MAX_RECURSION_DEPTH = 1000

def can_place_items(items, containers, h, index = 0, depth = 0)
  if depth > MAX_RECURSION_DEPTH
    puts "Превышена максимальная глубина рекурсии!"
    return false
  end

  return true if index == items.size

  containers.each_with_index do |container, i|
    if container + items[index] <= h
      containers[i] += items[index]
      if can_place_items(items, containers, h, index + 1, depth + 1)
        return true
      end
      containers[i] -= items[index] # Бэктрекинг
    end
  end

  false
end

# Ввод данных
print "Введите количество контейнеров (N): "
n = gets.to_i
print "Введите высоту каждого контейнера (H): "
h = gets.to_i
print "Введите количество предметов: "
m = gets.to_i

puts "Введите высоты предметов (через пробел): "
items = gets.split.map(&:to_i).sort.reverse # Сортируем по убыванию

containers = Array.new(n, 0)

if can_place_items(items, containers, h)
  puts "Предметы можно разместить в контейнерах!"
else
  puts "Предметы нельзя разместить в контейнерах."
end
